templates/verifier_groth16.vy.ejs

```diff
@@ -0,0 +1,189 @@
+# @version >=0.3.4
+
+# The prime q in the base field F_q for G1 
+BASE_MODULUS : constant(uint256) = 21888242871839275222246405745257275088696311157297823662689037894645226208583
+
+# The prime moludus of the scalar field of G1.
+SCALAR_MODULUS : constant(uint256) = 21888242871839275222246405745257275088548364400416034343698204186575808495617
+
+# the number of inputs
+N_INPUTS: constant(uint256) = <%=IC.length-1%> 
+
+struct G1Point:
+    X: uint256
+    Y: uint256
+
+# Encoding of field elements is: X[0] * z + X[1]
+struct G2Point:
+    X: uint256[2]
+    Y: uint256[2]
+
+struct VerifyingKey:
+    alfa1: G1Point
+    beta2: G2Point
+    gamma2: G2Point
+    delta2: G2Point
+    ic: DynArray[G1Point, <%=IC.length%>]
+
+struct Proof:
+    A: G1Point
+    B: G2Point
+    C: G1Point
+
+
+# @dev return the generator of G1
+@internal
+@pure
+def P1() -> G1Point:
+    return G1Point({X: 1, Y: 2})
+
+
+# @dev return the generator of G2
+@internal
+@pure
+def P2() -> G2Point:
+    return G2Point({
+        X: [
+          11559732032986387107991004021392285783925812861821192530917403151452391805634,
+          10857046999023057135944570762232829481370756359578518086990519993285655852781
+        ],
+        Y: [
+          4082367875863433681332203403145435568316851327593401208105741076214120093531,
+          8495653923123431417604973247489272438418190587263600148770280649306958101930
+        ]}
+      )
+
+
+# @dev return r the negation of p, i.e. p.addition(p.negate()) should be zero.
+@internal
+@view
+def negate(p: G1Point) -> G1Point:
+    if (p.X == 0) and (p.Y == 0):
+        return G1Point({X: 0, Y: 0})
+
+    # Validate input or revert
+    assert (p.X < BASE_MODULUS) or (p.Y < BASE_MODULUS), "invalid proof"
+
+    # We know p.Y > 0 and p.Y < BASE_MODULUS.
+    return G1Point({X: p.X, Y: BASE_MODULUS - p.Y})
```

Comment:
i'm not sure this logic matches the solidity implementation.

Reply:
I took this implementation from the semaphore repository, I feel it adds a layer of security since it prevents the use of invalid curve points. Here is the link to the implementation https://github.com/semaphore-protocol/semaphore/blob/26e0cbcc9dc474354c787b0294ea843efee42c67/packages/contracts/snarkjs-templates/verifier_groth16.sol.ejs#L64

Reply:
in that case - lgtm